service/controller/clusterapi/v17/resources/clusterstatus/create.go

```diff
@@ -2,8 +2,212 @@ package clusterstatus
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d nodes from tenant cluster", len(nodes)))
+		}
+	}
+
+	var machineDeployments []cmav1alpha1.MachineDeployment
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding MachineDeployments for tenant cluster")
+
+		l := metav1.AddLabelToSelector(
+			&v1.LabelSelector{},
+			label.Cluster,
+			key.ClusterID(&cr),
+		)
+		o := metav1.ListOptions{
+			LabelSelector: l.String(),
+		}
+
+		list, err := r.cmaClient.ClusterV1alpha1().MachineDeployments(cr.Namespace).List(o)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		machineDeployments = list.Items
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d MachineDeployments for tenant cluster", len(machineDeployments)))
+	}
+
+	updatedStatus := r.computeClusterConditions(ctx, cr, r.accessor.GetCommonClusterStatus(cr), nodes, machineDeployments)
+
+	if !reflect.DeepEqual(r.accessor.GetCommonClusterStatus(cr), updatedStatus) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "updating cluster status")
+
+		cr = r.accessor.SetCommonClusterStatus(cr, updatedStatus)
+
+		_, err = r.cmaClient.ClusterV1alpha1().Clusters(cr.Namespace).Update(&cr)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "updated cluster status")
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling reconciliation")
+		reconciliationcanceledcontext.SetCanceled(ctx)
+
+		return nil
+	}
+
 	return nil
 }
+
+func (r *Resource) computeClusterConditions(ctx context.Context, cluster cmav1alpha1.Cluster, clusterStatus v1alpha1.CommonClusterStatus, nodes []corev1.Node, machineDeployments []cmav1alpha1.MachineDeployment) v1alpha1.CommonClusterStatus {
+	var currentVersion string
+	var desiredVersion string
+	{
+		currentVersion = clusterStatus.LatestVersion()
+		desiredVersion = key.ReleaseVersion(&cluster)
+	}
+
+	// Count total number of all workers and number of Ready workers that
+	// belong to this cluster.
+	var desiredReplicas int
+	var readyReplicas int
+	{
+		for _, md := range machineDeployments {
+			desiredReplicas += int(md.Status.Replicas)
+		}
+
+		for _, md := range machineDeployments {
+			readyReplicas += int(md.Status.ReadyReplicas)
+		}
+	}
+
+	// After initialization the most likely implication is the tenant cluster
+	// being in a creation status. In case no other conditions are given and no
+	// versions are set, we set the tenant cluster status to a creating
+	// condition.
+	{
+		notCreating := !clusterStatus.HasCreatingCondition()
+		conditionsEmpty := len(clusterStatus.Conditions) == 0
+		versionsEmpty := len(clusterStatus.Versions) == 0
+
+		if notCreating && conditionsEmpty && versionsEmpty {
+			clusterStatus.Conditions = clusterStatus.WithCreatingCondition()
+			r.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("setting %#q status condition", v1alpha1.ClusterStatusConditionCreating))
+		}
+	}
```

Comment:
These are still fall-through-style as commented by Theo. I can live with these for now as this is how the earlier statusresource was as well. I think this is something that should be addressed as separate PR.

Reply:
As discussed. I agree with the unit tests. I do not agree on the different style. Will provide tests in a separate PR. This code was more or less copied from `statusresource` and works like that in production for almost a year already.